File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(strings, predicate)
  for i = 0, #strings do
    if strings[i] ~= nil and predicate(strings[i]) then
      return string.lower(strings[i])
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  local power = 1
  return coroutine.create(function()
    while power <= limit do
      coroutine.yield(power)
      power = power * base
    end
  end)
end


-- Write your say function here
function say(word)
  local str = {}

  function chain(next)
    if next ~= nil then
      table.insert(str, next)
      return chain
    else
      return table.concat(str, " ")
    end
  end

  if word == nil then
    return ""
  end
  table.insert(str, word)
  return chain
  
end


-- Write your line count function here
-- used http://lua-users.org/wiki/StringTrim to help me trim the whitespace
function meaningful_line_count(filename)
  local file = io.open(filename, "r")
  count = 0
  if file == nil then
    error("No such file")
  end
  
  for line in file:lines() do
    trimmedLine = line:gsub("^%s*(.-)%s*$", "%1") 

    if(trimmedLine ~= "" and not trimmedLine:match("^#")) then
      count = count + 1
    end
  end

  file:close()
  return count
end


Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a
  self.b = b
  self.c = c
  self.d = d
  return self
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion:__add(q)
  return Quaternion.new(
    self.a + q.a,
    self.b + q.b,
    self.c + q.c,
    self.d + q.d
  )
end

function Quaternion:__mul(q)
  return Quaternion.new(
    self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d,
    self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c,
    self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b,
    self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a
  )
end

function Quaternion:__eq(q)
  return self.a == q.a and self.b == q.b and self.c == q.c and self.d == q.d
end

function Quaternion:__tostring()
  local str = ""

  if self.a ~= 0 then
    str = tostring(self.a)
  end

  if self.b ~= 0 then
    if self.b == 1 then
      if str == "" then
        str = "i"
      else
        str = str .. "+i"
      end
    elseif self.b == -1 then
      str = str .. "-i"
    elseif self.b > 0 then
      str = str .. "+" .. tostring(self.b) .. "i"
    else 
      str = str .. tostring(self.b) .. "i"
    end
  end

  if self.c ~= 0 then
    if self.c == 1 then
      if str == "" then
        str = "j"
      else
        str = str .. "+j"
      end
    elseif self.c == -1 then
      str = str .. "-j"
    elseif self.c > 0 then
      str = str .. "+" .. tostring(self.c) .. "j"
    else 
      str = str .. tostring(self.c) .. "j"
    end
  end

  if self.d ~= 0 then
    if self.d == 1 then
      if str == "" then
        str = "k"
      else
        str = str .. "+k"
      end
    elseif self.d == -1 then
      str = str .. "-k"
    elseif self.d > 0 then
      str = str .. "+" .. tostring(self.d) .. "k"
    else 
      str = str .. tostring(self.d) .. "k"
    end
  end

  if str == "" then
    str = "0"
  end

  return str

end

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(strings, predicate, /):
    for string in strings:
        if predicate(string):
            return string.lower()
    return None


# Write your powers generator here
def powers_generator(base: int, limit: int):
    power = 1
    while power <= limit:
        yield power
        power = power * base
        
g = powers_generator(base=2, limit=7)
next(g)  
next(g)  
next(g)  
try:
    next(g)
except StopIteration:
    print("No more values")


# Write your say function here
def say(word = None):
    string = []
    def chain(next = None):
        if next is not None:
            string.append(next)
            return chain
        return ' '.join(string)
        
    if word is None:
        return chain()
    
    string.append(word)
    return chain


# Write your line count function here
# NOT empty, entirely whitepace, or whose first char is #

def meaningful_line_count(fileName = None):
    if fileName is not None:
        with open(fileName, 'r') as file:
            count = 0
            for line in file:
                if line.strip() and not line.strip().startswith('#'):
                    count = count + 1
        return count
    else:
        print("No such file")
        return FileNotFoundError
    
    
# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float 

    @property
    def conjugate(self) -> str:
        return Quaternion(self.a, (-1*self.b), (-1*self.c), (-1*self.d))
        #return None
    
    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)

    def __add__(self, q: "Quaternion"): # -> 'Quaternion':
        return Quaternion( 
            self.a + q.a,
            self.b + q.b,
            self.c + q.c,
            self.d + q.d,
        )
    
    def __mul__(self, q: "Quaternion"): # -> "Quaternion":

        a = self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d
        b = self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c
        c = self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b
        d = self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a

        return Quaternion(a, b, c, d)
    
    def __eq__(self, q: "Quaternion") -> bool:
        return (self.a, self.b, self.c, self.d) == (q.a, q.b, q.c, q.d)

    def __str__(self) -> str:
        string = ""
        if self.a != 0:
            string += f'{self.a}'

        if self.b != 0:
            if self.b == 1:
                if string == "":
                    string += "i"
                else:
                    string += "+i" 
            elif self.b == -1:
                string += "-i"
            elif self.b > 1:
                string += "+" + f'{self.b}i'
            else:
                string += f"{self.b}i"

        if self.c != 0:
            if self.c == 1:
                if string == "":
                    string += "j"
                else:
                    string += "+j" 
            elif self.c == -1:
                string += "-j"
            elif self.c > 1:
                string += "+" + f'{self.c}j'
            else:
                string += f"{self.c}j"

        if self.d != 0:
            if self.d == 1:
                if string == "":
                    string += "k"
                else:
                    string += "+k" 
            elif self.d == -1:
                string += "-k"
            elif self.d > 1:
                string += "+" + f'{self.d}k'
            else:
                string += f"{self.d}k"
        return string or "0"

    